src/utils/works.ts

```diff
@@ -7,16 +7,57 @@ import { getTagUrl } from "./tags";
 // correct type of page is passed to each method that extracts data.
 // Other than this, all pages are instances of CheerioAPI and can be used interchangeably.
 interface WorksPage extends CheerioAPI {
-  kind: "WorksPage";
+ kind: "WorksPage";
 }
+
+export interface WorkData {
+ workUrl: string;
+ workId: string;
+ chapterId: string | null;
+ collectionName: string | null;
+}
+
 const getWorksUrl = (tagName: string) => `${getTagUrl(tagName)}/works`;
 
 export const getWorksPage = async (tagName: string) => {
-  return cheerio.load(
-    (await axios.get<string>(getWorksUrl(tagName))).data
-  ) as WorksPage;
+ return cheerio.load(
+  (await axios.get<string>(getWorksUrl(tagName))).data
+ ) as WorksPage;
 };
 
 export const getTagId = ($worksPage: WorksPage) => {
-  return $worksPage(".rss")[0]?.attribs["href"].split("/")[2] || null;
+ return $worksPage(".rss")[0]?.attribs["href"].split("/")[2] || null;
+};
+
+interface WorkPage extends CheerioAPI {
+ kind: "WorkPage";
+}
+
+export const getWorkIdFromUrl = (url: string) => {
+ return url.match(/works\/(\d+)/)[1];
+};
+
+export const getChapterIdFromUrl = (url: string) => {
+ const chapterId = url.match(/chapters\/(\d+)/);
+ if (chapterId) return chapterId[1];
+};
+
+// ! this grabs the collection name from the URL but a work can be in more than one collection
+export const getCollectionNameFromUrl = (url: string) => {
+ const collectionName = url.match(/collections\/(\w+)/);
+ if (collectionName) return collectionName[1];
 };
+
+export const getWorkUrl = 
+({ workId, chapterId, collectionName }: {workId: string; chapterId?: string; collectionName?: string;}) => {
+if (chapterId && collectionName) {
```

Comment:
This is totally fine code, but here's a suggestion to make it a bit easier to follow:

```
let workUrl = "https://archiveofourown.org/";

if (collectionName) {
  workUrl += "collections/${collectionName}/";
}

workUrl += "works/${workId}/";

if (chapterId) {
  workUrl += "chapters/${chapterId}";
}
```

In this case the URL can be constructed bit by bit, and I think making it explicit to see what's added when variables are present is helpful cause it makes the structure of each URL explicit.